exporter/fileexporter/file_exporter.go

```diff
@@ -0,0 +1,207 @@
+	resourcepb "github.com/census-instrumentation/opencensus-proto/gen-go/resource/v1"
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/golang/protobuf/proto"
+
+	"github.com/open-telemetry/opentelemetry-collector/component"
+	"github.com/open-telemetry/opentelemetry-collector/consumer/consumerdata"
+)
+
+// Marshaler configuration used for marhsaling Protobuf to JSON. Use default config.
+var marshaler = &jsonpb.Marshaler{}
+
+// Helper struct to write JSON objects and arrays.
+type jsonWriter struct {
+	firstFieldDone     bool
+	firstArrayItemDone bool
+	writer             io.Writer
+}
+
+// Begin writing JSON. Call first.
+func (jw *jsonWriter) Begin() error {
+	_, err := io.WriteString(jw.writer, "{\n")
+	return err
+}
+
+// End writing JSON. Call last.
+func (jw *jsonWriter) End() error {
+	_, err := io.WriteString(jw.writer, "\n}\n")
+	return err
+}
+
+// MarshalObject marshals an object as a field of top-level object.
+func (jw *jsonWriter) MarshalObject(fieldName string, pb proto.Message) error {
+	if jw.firstFieldDone {
+		io.WriteString(jw.writer, ",\n")
+	} else {
+		jw.firstFieldDone = true
+	}
+	_, err := io.WriteString(jw.writer, `  "`+fieldName+`": `)
+	if err != nil {
+		return err
+	}
+
+	err = marshaler.Marshal(jw.writer, pb)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// BeginMarshalArray prepares to marshal array items under a field of top-level object.
+func (jw *jsonWriter) BeginMarshalArray(fieldName string) error {
+	if jw.firstFieldDone {
+		io.WriteString(jw.writer, ",\n")
+	} else {
+		jw.firstFieldDone = true
+	}
+	_, err := io.WriteString(jw.writer, `  "`+fieldName+"\": [")
+	jw.firstArrayItemDone = false
+	return err
+}
+
+// EndMarshalArray must be called after all array items are marshaled.
+func (jw *jsonWriter) EndMarshalArray() error {
+	var str string
+	if jw.firstArrayItemDone {
+		// Non-empty array. End on a new line.
+		str = "\n  ]"
+	} else {
+		// Empty array. End on the same line.
+		str = "]"
+	}
+	_, err := io.WriteString(jw.writer, str)
+	return err
+}
+
+// MarshalArrayItem marshals single array item. Call repeatedly after BeginMarshalArray.
+func (jw *jsonWriter) MarshalArrayItem(pb proto.Message) error {
+	var str string
+	if jw.firstArrayItemDone {
+		str = ",\n    "
+	} else {
+		str = "\n    "
+		jw.firstArrayItemDone = true
+	}
+	_, err := io.WriteString(jw.writer, str)
+	if err != nil {
+		return err
+	}
+	err = marshaler.Marshal(jw.writer, pb)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func exportResourceAndNode(writer *jsonWriter, node *commonpb.Node, resource *resourcepb.Resource) error {
+	if resource != nil {
+		err := writer.MarshalObject("resource", resource)
+		if err != nil {
+			return err
+		}
+	}
+	if node != nil {
+		return writer.MarshalObject("node", node)
+	}
+	return nil
+}
+
+// Exporter is the implementation of file exporter that writes telemetry data to a file
+// in Protobuf-JSON format.
+type Exporter struct {
```

Comment:
nit: in general I prefer the high-level stuff on top, personal preference...

Reply:
Fixed.

Reply:
I usually do the opposite: dependent code goes after what it depends on :-)